### PR TITLE
Audio/video segmentation and feature extraction

### LINKED ISSUE #5: segmentation
crop an audio file from faster-whisper timestamps (with pyannote)

### LINKED ISSUE #6: feature extraction
use wavLM for feature extraction from audio

### LINKED ISSUE #8: video segmentation
implement video segmentation for multi-modal emotion recognition
ffmpeg

### PR DESCRIPTION
**Pull request by Martaesplo**
#5 #8 For segmenting an audio and/or video file using whisper's timestamps: segment.py
#6 For generating speaker embeddings for each given audio segment: embed.py
To run this two steps of the pipeline: run_faster_whisper.py

Example outputs of the segmentation can be seen on the two existing folders, only performed for one segment.

The rest of the files can be ignored as they are old, or needed modules.

### CODE REVIEW

File: __pycache__/WavLM.cpython-310.pyc
Comment (greenw0lf) [inline __pycache__/WavLM.cpython-310.pyc:1]: You should create a `.gitignore` file which has `__pycache__` as a line inside it. `__pycache__` is not something you want to have in a repository.

File: WavLM.py
Comment (greenw0lf) [inline WavLM.py:1]: Is this a modified version of WavLM? If not, you should look into adding it as a dependency (it's a similar issue with the fairseq folder that Kathleen wanted to add to her repo)

File: embed.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+#decode audio with faster-whisper:
+#sr1 = 16000
+#waveform1 = decode_audio(audio_name, sr1)
+
+def load_model():
+  '''
+  Loads feature extraction model, configuration and weights.
+  '''
+  # load the pre-trained checkpoints
+  model_checkpoint_loc = 'WavLM-Base+.pt' #WavLM-Large consumed way too much RAM
```
Comment (greenw0lf) [inline embed.py:11]: This code will be running at some point in a DANE worker, right? I would imagine there will be more resources available to run a larger version of WavLM, so it would make sense to change the checkpoint to `WavLM-Large`. Maybe Sara can comment more about this.

File: embed.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+from WavLM import WavLM, WavLMConfig
+
+#audio_name = "testaudio.wav"
+
+#decode audio with librosa:
+#waveform1, sr = librosa.load(audio_name)
+
+#decode audio with faster-whisper:
+#sr1 = 16000
+#waveform1 = decode_audio(audio_name, sr1)
```
Comment (greenw0lf) [inline embed.py]: If this commented codeblock will not be used, I suggest removing it

File: embed.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+def embed_audio(model, cfg, audio_segment, last_layer=False):
+  '''
+  Generates embeddings for the given audio file or segment. Extracts the
+  representation of the last layer (last_layer=True), or the representation from
+  each layer and performs a weighted sum (last_layer=False).
+  '''
+
+  #if last_layer:
+  #if cfg.normalize:
```
Comment (greenw0lf) [inline embed.py:27]: Same comment as before, if not (to be) used -> remove

Comment (greenw0lf) [reply embed.py:27]: If used, you should add an extra comment saying this will be uncommented or something.

**Comment by greenw0lf** [comment]
You should link the PR to the issues that it addresses (you can do so by typing # and then you will get a list of issues to choose from)

File: run_faster_whisper.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+from faster_whisper import WhisperModel
+from segment import segment_audio, segment_video
+from embed import load_model, embed_audio
+#from embed import embed_audio
```
Comment (greenw0lf) [inline run_faster_whisper.py]: If not used (later on) -> remove

File: run_faster_whisper.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+#REMINDER: using the .local faster_whisper
+
+model_size = "large-v3"
+
+# Run on GPU with FP16
+#model = WhisperModel(model_size, device="cuda", compute_type="float16")
+
+# or run on GPU with INT8
+# model = WhisperModel(model_size, device="cuda", compute_type="int8_float16")
```
Comment (greenw0lf) [inline run_faster_whisper.py:14]: You should probably look into using these (going back to the more resources available thing I mentioned in a different comment)

File: run_faster_whisper.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+from embed import load_model, embed_audio
+#from embed import embed_audio
+import tensorflow as tf
+import torchaudio, torch
+import os
+from WavLM import WavLM, WavLMConfig
+
+#REMINDER: using the .local faster_whisper
+
+model_size = "large-v3"
```
Comment (greenw0lf) [inline run_faster_whisper.py]: Speaking from experience, v3 performs slightly worse than v2 (in all cases that I evaluated on for my project it was worse).

File: marta_video_segments/marta_00:00:07_00:00:15.mp4
Comment (greenw0lf) [inline marta_video_segments/marta_00:00:07_00:00:15.mp4:1]: Input/output files usually shouldn't be added to GitHub. Rather, whoever wants to test out your code can request them from you. Maybe you can keep them for now until we finish testing your code, then remove them.

File: segment.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+#from faster_whisper import WhisperModel
+#from faster_whisper.audio import decode_audio
```
Comment (greenw0lf) [inline segment.py]:
```suggestion

```

File: segment.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+#import librosa
+
+#DEPENDENCIES: faster-whisper, pyannote.audio, pyannote.core 
+
+#decode audio with librosa:
+#waveform1, sr = librosa.load(audio_name)
+
+#decode audio with faster-whisper:
+#sr1 = 16000
+#waveform1 = decode_audio(audio_name, sr1)
```
Comment (greenw0lf) [inline segment.py]:
```suggestion

```

File: segment_embed.py
Comment (greenw0lf) [inline segment_embed.py:1]: Is this file necessary? It seems to be a mashup of `segment.py` and `run_faster_whisper.py`

**Review by greenw0lf — CHANGES_REQUESTED** [review]
Left some comments about what is inside the code. I will be testing whether it works or not to also confirm the functionality of it.

**Comment by greenw0lf** [comment]
I can't pull your branch because of files that contain `:` in the name. I suggest replacing `:` with `-` to ensure compatibility with all OS's

File: WavLM.py
Comment (Veldhoen) [reply WavLM.py:1]: Agree with Dragos! Ask me (or Kathleen) if you need help to asjust this

File: __pycache__/WavLM.cpython-310.pyc
Comment (Veldhoen) [reply __pycache__/WavLM.cpython-310.pyc:1]: Indeed, add .gitignore. See https://github.com/beeldengeluid/dane-example-worker/blob/main/.gitignore for reference (I only just realized I had not added it yet :$)

File: embed.py (hunk 1 above)
Comment (Veldhoen) [reply embed.py:11]: Yes, I might be nice to make this a configurable option. Ideally, we would do some tests to see the impact of this too. 

In any case, we need to configure all locations rather than hardcode them, in the end.

File: embed.py [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+  wav_input_16khz = torch.nn.functional.layer_norm(audio_segment, audio_segment.shape)
+  rep = model.extract_features(wav_input_16khz)[0]
+
+  if last_layer == True:
+    if cfg.normalize:
+      wav_input_16khz = torch.nn.functional.layer_norm(wav_input_16khz , wav_input_16khz.shape)
+    rep, layer_results = model.extract_features(wav_input_16khz,
+                                                output_layer=model.cfg.encoder_layers,
+                                                ret_layer_results=True)[0]
+    layer_reps = [x.transpose(0, 1) for x, _ in layer_results]
```
Comment (Veldhoen) [inline embed.py:37]: layer_reps is never used or passed to any other function, right? So ret_layer_results might as well be set to False then?

File: embed.py
Comment (Veldhoen) [inline embed.py:1]: Please adhere to code style: apply black and flake8 with these settings: https://github.com/beeldengeluid/dane-example-worker/blob/main/.flake8 (you can copy the .flake8 to this repo)

File: marta_video_segments/marta_00:00:07_00:00:15.mp4
Comment (Veldhoen) [reply marta_video_segments/marta_00:00:07_00:00:15.mp4:1]: Agree. We sometimes add some small sample data to tests/data though, if it is needed to run unit tests. Please consider setting that up!

File: run_faster_whisper.py (hunk 5 above)
Comment (Veldhoen) [reply run_faster_whisper.py:14]: Yes, infer appropriate compute from context or make it configurable

File: run_faster_whisper.py
Comment (Veldhoen) [inline run_faster_whisper.py:1]: This is still quite rough code that needs some refactoring into clean functions.

File: modules.py
Comment (Veldhoen) [inline modules.py:1]: Use a different construct to include WavLM than just injecting the files here. I'd be happy to help!

File: segment.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  vend = seconds_to_hms(trans_segment.end)
+
+  #extracting filename and extension
+  name, extension = os.path.splitext(video)
+  if not os.path.exists("%s_video_segments" % (name)):
+      os.makedirs("%s_video_segments" % (name))
+  
+  outfile = "./%s_video_segments/%s_%s_%s%s" % (name, name, vstart, vend, extension)
+  
+  run_ffmpeg = ['ffmpeg', '-ss', vstart, '-to', vend, '-i', video, '-c', 'copy', outfile]
```
Comment (Veldhoen) [inline segment.py]: It might be worthwile to create one input stream for ffmpeg and then output segments from there, rather than calling ffmpeg for each segment individually.

File: segment_embed.ipynb
Comment (Veldhoen) [inline segment_embed.ipynb:1]: I assume this is notmeant to go into the repository, right?

File: testaudio_out.txt
Comment (Veldhoen) [inline testaudio_out.txt:1]: Only add (sample) output if it is necessary for testing (and in that case, put it in an appropriate directory)

**Review by Veldhoen — CHANGES_REQUESTED** [review]
Congrats on pushing all your code to GH and creating your first PR!
@greenw0lf raised a few good points and I added some suggestions too, here and there. Please take a look and don't hesitate to ask for help!
